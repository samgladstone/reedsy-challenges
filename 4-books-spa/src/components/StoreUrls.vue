<template>
    <div>
        <a v-for="(url, store) in book.stores"
           :key="store"
           :href="url">{{store}}</a>
        <span>Not Available</span>
    </div>
</template>

<script>
    import { mapState } from 'vuex';

    export default {
        name: 'StoreUrls',
        props: {
            bookId: {
                type: Number,
                required: true,
            },
        },
        computed: {
            ...mapState({
                book(state) {
                    return state.books[this.bookId];
                },
            }),
        },
    };
</script>

<style lang="scss" scoped>
    a {
        display: inline;
        color: #c0d58c;
        text-decoration: none;
        font-weight: 600;
        white-space: nowrap;

        &:not(:last-child) {
            padding-right: 1em;
        }
    }

    @media (min-width: $urlsBreakpoint) {
        a:not(:last-child) {
            display: block;
            padding-right: 0;
        }
    }

    span {
        font-style: italic;
        color: #aaa;
        display: none;

        &:first-child {
            display: block;
        }
    }
</style>