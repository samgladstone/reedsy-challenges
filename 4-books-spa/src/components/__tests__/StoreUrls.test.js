import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import StoreUrls from '../StoreUrls.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

let store, mockBook;

beforeEach(() => {
    mockBook = {
        "id": 1,
        "author": "Markus Zusak",
        "cover": "11.jpg",
        "rating": "9.9",
        "published": "2005",
        "slug": "the-book-theif",
        "synopsis": "1939. Nazi Germany. The country is holding its breath. Death has never been busier. Liesel, a nine-year-old girl, is living with a foster family on Himmel Street. Her parents have been taken away to a concentration camp. Liesel steals books. This is her story and the story of the inhabitants of her street when the bombs begin to fall.",
        "title": "The Book Thief",
        "stores": {
            "Amazon": "https://www.Amazon.co.uk/Book-Thief-Markus-Zusak/dp/0552773891",
            "Apple Books": "https://books.apple.com/gb/book/the-book-theif/id1078821147",
            "Play Store": "https://play.google.com/store/audiobooks/details/Markus_Zusak_The_Book_Thief?id=AQAAAEDgxmqGmM&hl=en"
        }
    };

    store = new Vuex.Store({
        state: {
            books: {
                1: mockBook
            }
        }
    })
})

const factory = (propsData = { bookId: 1 }) => {
    return shallowMount(StoreUrls, {
        store, localVue,
        propsData
    });
}

test('The correct book is looked up', () => {
    const book2 = store.state.books[2] = { ...mockBook, id: 2 };

    expect(factory().vm.book).toBe(mockBook);
    expect(factory({ bookId: 2 }).vm.book).toBe(book2);
});

describe('Link rendering', () => {
    test('No stores', () => {
        delete store.state.books[1].stores;
        expect(factory()).toMatchSnapshot();
    });

    test('Empty stores', () => {
        store.state.books[1].stores = {};
        expect(factory()).toMatchSnapshot();
    });

    test('One store', () => {
        const stores = store.state.books[1].stores;
        delete stores.Amazon;
        delete stores['Play Store'];
        expect(factory()).toMatchSnapshot();
    });

    test('All three stores', () => {
        expect(factory()).toMatchSnapshot();
    });
});