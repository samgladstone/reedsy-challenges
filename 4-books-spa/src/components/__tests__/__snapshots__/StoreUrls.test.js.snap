// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link rendering All three stores 1`] = `<div><a href="https://www.Amazon.co.uk/Book-Thief-Markus-Zusak/dp/0552773891">Amazon</a><a href="https://books.apple.com/gb/book/the-book-theif/id1078821147">Apple Books</a><a href="https://play.google.com/store/audiobooks/details/Markus_Zusak_The_Book_Thief?id=AQAAAEDgxmqGmM&amp;hl=en">Play Store</a> <span>Not Available</span></div>`;

exports[`Link rendering Empty stores 1`] = `<div> <span>Not Available</span></div>`;

exports[`Link rendering No stores 1`] = `<div> <span>Not Available</span></div>`;

exports[`Link rendering One store 1`] = `<div><a href="https://books.apple.com/gb/book/the-book-theif/id1078821147">Apple Books</a> <span>Not Available</span></div>`;
